### PR TITLE
kv: track withCommit across concurrent batches split by divideAndSendBatchToRanges

### PR DESCRIPTION
**Pull request by nvb**
~If a batch is performing a parallel commit and the transaction has previously pipelined writes that have yet to be proven successful then the EndTransaction request will be preceded by a series of QueryIntent requests (see txn_pipeliner.go). Before evaluating, each of these QueryIntent requests will grab latches and wait for their corresponding write to finish. This is how the QueryIntent requests synchronize with the write they are trying to verify.~

~If these QueryIntents remained in the same batch as the EndTransaction request then they would force the EndTransaction request to wait for the previous write before evaluating itself. This "pipeline stall" would effectively negate the benefit of the parallel commit. To avoid this, the final commit makes sure that these "pre-commit" QueryIntent requests are split from and issued concurrently with the rest of the parallel commit batch.~

~With this change in place, parallel commits work and demonstrate their expected performance win.~

This commit puts in place the logic to return ambiguous errors not only by batches containing EndTransaction(commit=false) requests, but also by batches sent concurrently with EndTransaction(commit=false) requests.

This doesn't change any behavior immediately because divideAndSendBatchToRanges will currently return an `errNo1PCTxn` if if splits a batch with an EndTransaction request.

### CODE REVIEW

**Review by tbg — APPROVED** [review]
LGTM for all but last commit, which I haven't reviewed in anticipation of its cleanup. I think it's worth merging the rest first! The post-rebase reviews are already pretty annoying.

> Reviewed 1 of 10 files at r2, 1 of 3 files at r4, 8 of 12 files at r6, 26 of 29 files at r7, 2 of 2 files at r8, 1 of 1 files at r9, 3 of 3 files at r10, 17 of 17 files at r11, 2 of 2 files at r12, 1 of 1 files at r13, 3 of 3 files at r14, 25 of 25 files at r16, 2 of 2 files at r17, 1 of 1 files at r18, 3 of 3 files at r19, 6 of 6 files at r20.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-:-Lf2GX5v-j6id0O2PXep:b-9xiuv3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/dist_sender.go, line 311 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0e4koB7ttPRtOfGbT:-Lf0e4kp1qD-G64ho6DT:b86juuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a406cb843e494f035623222fca38f3d4cb2777/pkg/kv/dist_sender.go#L311)):*
> ```Go
> }
> 
> // DisableParallelBatches disableParallelBatches instructs DistSender to never
> ```

copy pasta
___
*[pkg/kv/dist_sender.go, line 809 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0eWIB0kpJ9fXcemNI:-Lf0eWIB0kpJ9fXcemNJ:b-nccjqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cc0eb393dd2d2b663dd6b71d11947713f9933b5/pkg/kv/dist_sender.go#L809)):*
> ```Go
> 	}
> 
> 	// The batch spans ranges. Verify that this is ok.
> ```

I don't know if this is worth pointing out, but it's not proven that the batch spans ranges. The ranges could've merged, which is a TODO in my name further down. (The TODO would be better kept around here).
___
*[pkg/kv/dist_sender.go, line 633 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0fPwkDfZ89nnydHIu:-Lf0fPwkDfZ89nnydHIv:b-qd5hv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1b24816d68a92778fa4c2b9608f5292d56bea1/pkg/kv/dist_sender.go#L633)):*
> ```Go
> 	if et, ok := lastReq.(*roachpb.EndTransactionRequest); ok && et.NoRefreshSpans {
> 		hasRefreshSpans := false
> 	Loop:
> ```

I hate labels, if you're on board with that you could

```go
hasRefreshSpans := func() bool {
  for { for { if stuff() { return false} } }
  return true
}()
```

but I'll live if you'd rather not.
___
*[pkg/kv/dist_sender.go, line 1364 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0hR1d1S-uwVvilZn6:-Lf0hR1d1S-uwVvilZn7:beppemv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1b24816d68a92778fa4c2b9608f5292d56bea1/pkg/kv/dist_sender.go#L1364)):*
Maybe rephrase this to say something along the lines of

> If withCommit is false (i.e. either no EndTransaction is in flight, or it is attempting to abort), ambiguous results will never be returned from this method. This is because both transactional writes and aborts can be retried (the former due to seqno idempotency, the latter because aborting is idempotent).

Question: can't continuing past an "uncertain abort" give you weird outcomes? Aborting an already aborting transaction, won't that give you a transaction status error or the like? (If so seems that we might want to improve on the behavior there). In any case, I like what you're going after here.
___
*[pkg/kv/dist_sender.go, line 780 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0pZas30clRu8ws0C2:-Lf0pZas30clRu8ws0C3:b2dp5m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a35ca567fd1c81b06be534af24b407270d83c29f/pkg/kv/dist_sender.go#L780)):*
> , adjusting the error index, if necessary.
___
*[pkg/kv/dist_sender.go, line 962 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0wyMp6uYOyBWruzCU:-Lf0wyMp6uYOyBWruzCV:b-gmv3m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a35ca567fd1c81b06be534af24b407270d83c29f/pkg/kv/dist_sender.go#L962)):*
> ```Go
> 	// previously pipelined writes that have yet to be proven successful then
> 	// the EndTransaction request will be preceded by a series of QueryIntent
> 	// requests (see txn_pipeliner.go). Before evaluating, each of these
> ```

Can there be any other QueryIntent in the batch other than the block right before EndTransaction? I'm thinkin of examples in which a key is written to in the last two batches of the txn. In that case I suspect we might be chaining on to it right before the write *and* before the endtxn, and if we don't catch both we may pipeline stall the endtxn again.

This is really more a question about what guarantees the txn pipeliner makes about where it places the queryintents, and whether it's smart enough to not chain to writes more than once in the same batch. (Or whether that's maybe just not something we care about).

By the way, we should be able to verify that these stalls don't happen. We basically never want QueryIntent and EndTransaction(commit=true) in the same batch on a range. I'm sure we can verify this somewhere in DistSender too, there are various places that loop over everything already anyway.

**Review by nvb — COMMENTED** [review]
TFTR!

I pulled the last commit out of this and plan to turn it into a new PR.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-:-Lf6ByqJ1b44WmNjTIQA:b98xfwj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/dist_sender.go, line 311 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0e4koB7ttPRtOfGbT:-Lf2NfOc2jISWpBlPaPS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a406cb843e494f035623222fca38f3d4cb2777/pkg/kv/dist_sender.go#L311)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

copy pasta
</blockquote></details>

Done.
___
*[pkg/kv/dist_sender.go, line 809 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0eWIB0kpJ9fXcemNI:-Lf2O7ORAIwcmfImxQ18:b-v0vx5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cc0eb393dd2d2b663dd6b71d11947713f9933b5/pkg/kv/dist_sender.go#L809)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I don't know if this is worth pointing out, but it's not proven that the batch spans ranges. The ranges could've merged, which is a TODO in my name further down. (The TODO would be better kept around here).
</blockquote></details>

Commented and moved the TODO up.
___
*[pkg/kv/dist_sender.go, line 633 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0fPwkDfZ89nnydHIu:-Lf2Od3uEXjRNWRfhYgF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1b24816d68a92778fa4c2b9608f5292d56bea1/pkg/kv/dist_sender.go#L633)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I hate labels, if you're on board with that you could

```go
hasRefreshSpans := func() bool {
  for { for { if stuff() { return false} } }
  return true
}()
```

but I'll live if you'd rather not.
</blockquote></details>

Done.
___
*[pkg/kv/dist_sender.go, line 1364 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0hR1d1S-uwVvilZn6:-Lf2qSWq-sP_yRzHUwos:bmhxjgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c1b24816d68a92778fa4c2b9608f5292d56bea1/pkg/kv/dist_sender.go#L1364)):*
> Maybe rephrase this to say something along the lines of

Done.

> Aborting an already aborting transaction, won't that give you a transaction status error or the like?

No, it doesn't look like it from [here](https://github.com/cockroachdb/cockroach/blob/7b2651400b2003d0a381cba9dbfc0b7bc0dfee00/pkg/storage/batcheval/cmd_end_transaction.go#L214). EndTransaction(commit=false) is idempotent.
___
*[pkg/kv/dist_sender.go, line 780 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0pZas30clRu8ws0C2:-Lf2u6zoAsXNj1cxXWOY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a35ca567fd1c81b06be534af24b407270d83c29f/pkg/kv/dist_sender.go#L780)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> , adjusting the error index, if necessary.
</blockquote></details>

Done.
___
*[pkg/kv/dist_sender.go, line 962 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/37469#-Lf0wyMp6uYOyBWruzCU:-Lf2rnj8A7K3UBAnkwQc:brdj2ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/a35ca567fd1c81b06be534af24b407270d83c29f/pkg/kv/dist_sender.go#L962)):*
> In that case I suspect we might be chaining on to it right before the write and before the endtxn

We won't chain on to the same key twice (see the [`txnPipeliner`](https://github.com/cockroachdb/cockroach/blob/7b2651400b2003d0a381cba9dbfc0b7bc0dfee00/pkg/kv/txn_interceptor_pipeliner.go#L397)), but it is absolutely possible that a request in the final batch overlaps a previous pipelined write and is prefixed with a QueryIntent as a result. This works as intended in that it will never miss a previous write but will induce a pipeline stall. This ends up being no different than any other pipeline stall in the transaction.

> This is really more a question about what guarantees the txn pipeliner makes about where it places the queryintents, and whether it's smart enough to not chain to writes more than once in the same batch. (Or whether that's maybe just not something we care about).

Yeah, the link I posted above points at the logic that ensures this. It's also logic that will be improved with the unsafe `ImmutableBytesSet` patch I have sitting around.

> By the way, we should be able to verify that these stalls don't happen. We basically never want QueryIntent and EndTransaction(commit=true) in the same batch on a range. I'm sure we can verify this somewhere in DistSender too, there are various places that loop over everything already anyway.

Asserting that somewhere is a good idea. The condition is actually a bit more complex because a non-parallel commit in the same batch as a QueryIntent is fine, but that's probably not worth supporting.

This brings up an interesting case that isn't correctly handled by this commit which I'll need to fix. Right now the QueryIntents won't be split from a parallel committing EndTransaction if the final batch doesn't span range boundaries. That's not the behavior we want. FWIW the more explicit tiered batch splitting that we were talking about today would have handled this correctly.
